blazar/plugins/devices/k3s_plugin.py

```diff
@@ -0,0 +1,184 @@
+
+        if device_name is None:
+            raise manager_ex.InvalidHost(host=device_values)
+
+        node = self.get_node_by_name(device_name)
+        if not node:
+            raise manager_ex.DeviceNotFound(device=device_name)
+
+        device_properties = {
+            'name': device_name,
+            'device_type': 'container',
+            'device_driver': K3sPlugin.device_driver
+        }
+
+        to_store = set(device_values.keys()) - set(device_properties.keys())
+        extra_capabilities_keys = to_store
+        extra_capabilities = dict(
+            (key, device_values[key]) for key in extra_capabilities_keys
+        )
+
+        if any([len(key) > 64 for key in extra_capabilities_keys]):
+            raise manager_ex.ExtraCapabilityTooLong()
+
+        device = None
+        cantaddextracapability = []
+        try:
+            device = db_api.device_create(device_properties)
+        except db_ex.BlazarDBException:
+            raise
+        for key in extra_capabilities:
+            values = {'device_id': device['id'],
+                      'capability_name': key,
+                      'capability_value': extra_capabilities[key],
+                      }
+            try:
+                db_api.device_extra_capability_create(values)
+            except db_ex.BlazarDBException:
+                cantaddextracapability.append(key)
+        if cantaddextracapability:
+            raise manager_ex.CantAddExtraCapability(
+                keys=cantaddextracapability,
+                host=device['id'])
+        self.set_device(device_name, device['id'])
+        return device['id']
+
+    def is_active(self, node):
+        conditions = node.status.conditions
+        for i in conditions:
+            if i.type == "Ready":
+                return i.status == "True"
+        return False
+
+    def poll_resource_failures(self, devices):
+        failed_devices = []
+        recovered_devices = []
+
+        for device in devices:
+            found_node = False
+            for node in self.get_nodes_by_label(LABELS["device"], None):
+                if device["name"] == node.metadata.name:
+                    found_node = True
+                    if not self.is_active(node) and device["reservable"]:
+                        failed_devices.append(device)
+                    elif self.is_active(node) and not device["reservable"]:
+                        recovered_devices.append(device)
+                    break
+            if not found_node:
+                failed_devices.append(device)
+
+        return failed_devices, recovered_devices
+
+    def allocate(self, device_reservation, lease, devices):
+        for device in devices:
+            self.set_res_id_label(
+                device["name"], device_reservation["reservation_id"])
+
+    def deallocate(self, device_reservation, lease, devices):
+        namespace = lease["project_id"]
+        for device in devices:
```

Comment:
Does setting a label to None remove it effectively?

Reply:
Yes, this clears the label.